user_ldap: Add support for gidNumber

Xuanwo opened the pull request:
This patch is based on the work of @dleeuw (https://github.com/dleeuw)
(See https://github.com/nextcloud/server/issues/2640#issuecomment-269615883 for more details).
The difference is user & group data will be written into cache to have
better performance, and functions splited from primaryGroupID series to
make them more readable.

Signed-off-by: Xuanwo <xuanwo@yunify.com>

Fixed https://github.com/nextcloud/server/issues/2640

MorrisJobke commented:
@blizzz Any comments on this?

Xuanwo commented:
@MorrisJobke Hi, what else should I do to make this pr merged into master?

MorrisJobke commented:
> @MorrisJobke Hi, what else should I do to make this pr merged into master?

Find someone to review it - and I would like @blizzz to do so, because he is highly into the LDAP code 😉

blizzz commented:
@Xuanwo nice work, thanks! I had a brief look and it looks good, I will do some further testing today or tomorrow.

Xuanwo commented:
@blizzz Sorry for disturbing again and again, but I want to know which stage my pr is on ?

Xuanwo commented:
any update?

Xuanwo commented:
@blizzz That's great, waiting for your review.

blizzz commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -229,9 +230,9 @@ private function _groupMembers($dnGroup, &$seen = null) {
 			}
 		}
-		
+
```
unnecessary white spaces (also below)

Xuanwo replied:
Yes, and my pr will remove them.

blizzz commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -263,7 +264,167 @@ private function _getGroupDNsFromMemberOf($DN, &$seen = null) {
+	 * @return string|bool
+	 */
+	public function gidNumber2Name($gid, $dn) {
+		$cacheKey = 'gidNumbertoName';
```
ideally, the T would be capitalized

Xuanwo replied:
Got it.

Xuanwo replied:
I do it this way because in primaryGroupID2Name function, the cache name is `primaryGroupIDtoName`

blizzz replied:
That naming is unfortunate too, should have spelled it "primaryGroupIdToName" back then. I shall wear sackcloth and ashes! :)

blizzz commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -263,7 +264,167 @@ private function _getGroupDNsFromMemberOf($DN, &$seen = null) {
+		}
+
+		//we need to get the DN from LDAP
+		$filter = $this->access->combineFilterWithAnd(array(
```
instead of ``array()`` we use the short form to create arrays ``[ ]`` meanwhile

Xuanwo replied:
OK

Xuanwo replied:
Other part of LDAP still use `array()`, should they also need tobe changed into `[]` ?

blizzz replied:
Old-style is used in older code, because older PHP versions were supported back then. When touching code they can be converted, too, but at least new code should follow the short notation.

blizzz commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -263,7 +264,167 @@ private function _getGroupDNsFromMemberOf($DN, &$seen = null) {
+		//we need to get the DN from LDAP
+		$filter = $this->access->combineFilterWithAnd(array(
+			$this->access->connection->ldapGroupFilter,
+			'objectClass=posixGroup',
```
I do not think it should be hardcoded. All constraints should be within the ldapGroupFilter. I.e. this line could just go imho. What do you think?

Xuanwo replied:
Let's find a better way to do it.

Xuanwo replied:
I'm not sure I have done it well, but from my point of view, `ldapGroupFilter` is config that user input, not a place for constraints.

blizzz replied:
These are admin-defined constraints. We do not need to narrow it down here, because either the attribute is present and matches, or not. If someone creates and uses a "reinventedPosixGroup" objectclass that permits a gidNumber, the feature would not work. Relying on the attribute is sufficient imho.

blizzz commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -263,7 +264,167 @@ private function _getGroupDNsFromMemberOf($DN, &$seen = null) {
+			$filter = $this->prepareFilterForUsersHasGidNumber($groupDN, $search);
+			$users = $this->access->fetchListOfUsers(
+				$filter,
+				array($this->access->connection->ldapUserDisplayName, 'dn'),
```
you can also use short notation here

Xuanwo replied:
Done

blizzz commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -263,7 +264,167 @@ private function _getGroupDNsFromMemberOf($DN, &$seen = null) {
+	public function countUsersInGidNumber($groupDN, $search = '', $limit = -1, $offset = 0) {
+		try {
+			$filter = $this->prepareFilterForUsersHasGidNumber($groupDN, $search);
+			$users = $this->access->countUsers($filter, array('dn'), $limit, $offset);
```
and short array notation here

Xuanwo replied:
Done.

blizzz commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -263,7 +264,167 @@ private function _getGroupDNsFromMemberOf($DN, &$seen = null) {
+	 * @param string $dn
+	 * @return string|bool
+	 */
+	public function getUserGidNumbers($dn) {
```
the plural form of the method name is irritating, also because of the similarity to its callee ``getUserGidNumber``, but mainly because it fetchesa single number. Alternatives perhaps: ``getUserGidNumberFromLDAP`` or ``readUserGidNumber``.

blizzz replied:
Or better make it singular and adjust the other method name (see below)

Xuanwo replied:
I changed `getUserGidNumber` into `getUserGroupByGid`, so that we can use `getUserGidNumber` here.

blizzz replied:
thx!

blizzz commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -263,7 +264,167 @@ private function _getGroupDNsFromMemberOf($DN, &$seen = null) {
+	 * @param string $dn
+	 * @return string
+	 */
+	public function getUserGidNumber($dn) {
```
actually, it gets the name, not the number. Maybe rename to ``getUserGroupByGid``.

Xuanwo replied:
Done.

blizzz commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -457,6 +618,7 @@ public function getUserGroups($uid) {
 
 		$groups = [];
 		$primaryGroup = $this->getUserPrimaryGroup($userDN);
+		$gidNumber = $this->getUserGidNumber($userDN);
```
it's not a number but a groupname, so perhaps ``$gidGroupName`` to make it more clear.

Xuanwo replied:
Done.

blizzz reviewed (commented):
Nice code and awesome that it comes with unit tests 🚀 

One thing could be spared: the hard coded filter part

Two methods and a variable name are confusing, they should be adjusted for better understanding.

Other comments are just formalities that would not be blockers, but could be fixed by the side.

Xuanwo committed:
user_ldap: Add support for gidNumber

This patch is based on the work of @dleeuw (https://github.com/dleeuw)
(See https://github.com/nextcloud/server/issues/2640#issuecomment-269615883 for more details).
The difference is user & group data will be written into cache to have
better performance, and functions splited from primaryGroupID series to
make them more readable.

Fixed https://github.com/nextcloud/server/issues/2640

Signed-off-by: Xuanwo <xuanwo@yunify.com>

blizzz commented:
Sign-off check fails for formal reasons, it expects the email address within ``< >``  brackets 🤷

Xuanwo commented:
@blizzz Should be fixed.

blizzz commented:
@Xuanwo thanks, looks good so far! One thing I found when testing: selecting a group on the users page won't load the members that have it as primary group via gidnumber. Can you confirm?

MorrisJobke commented:
> @Xuanwo thanks, looks good so far! One thing I found when testing: selecting a group on the users page won't load the members that have it as primary group via gidnumber. Can you confirm?

@Xuanwo Any news on this?

Xuanwo commented:
@MorrisJobke Sorry for late, I will have a test on the problem @blizzz reported.

Xuanwo commented:
@MorrisJobke @blizzz I fixed that problem, please review again. I test this patch on my server, it works well.

blizzz commented:
Works! Needs a changed method name call due to changes on master, I'll quickly commit and open a new PR, so we can finally merge it.

blizzz commented:
→ https://github.com/nextcloud/server/pull/4489
